kotsadm/pkg/socketservice/socketservice.go

```diff
@@ -0,0 +1,543 @@
+
+type ClusterSocket struct {
+	ClusterID             string
+	SocketID              string
+	SentPreflightURLs     map[string]bool
+	LastDeployedSequences map[string]int64
+}
+
+type DeployArgs struct {
+	AppID                string   `json:"app_id"`
+	AppSlug              string   `json:"app_slug"`
+	KubectlVersion       string   `json:"kubectl_version"`
+	AdditionalNamespaces []string `json:"additional_namespaces"`
+	ImagePullSecret      string   `json:"image_pull_secret"`
+	Namespace            string   `json:"namespace"`
+	PreviousManifests    string   `json:"previous_manifests"`
+	Manifests            string   `json:"manifests"`
+	Wait                 bool     `json:"wait"`
+	ResultCallback       string   `json:"result_callback"`
+	ClearNamespaces      []string `json:"clear_namespaces"`
+	ClearPVCs            bool     `json:"clear_pvcs"`
+	AnnotateSlug         bool     `json:"annotate_slug"`
+}
+
+type AppInformersArgs struct {
+	AppID     string   `json:"app_id"`
+	Informers []string `json:"informers"`
+}
+
+type SupportBundleArgs struct {
+	URI string `json:"uri"`
+}
+
+var server *socket.Server
+var clusterSocketHistory = []ClusterSocket{}
+var socketMtx sync.Mutex
+
+func Start() *socket.Server {
+	server = socket.NewServer(transport.GetDefaultWebsocketTransport())
+
+	server.On(socket.OnConnection, func(c *socket.Channel, args interface{}) {
+		socketMtx.Lock()
+		defer socketMtx.Unlock()
+
+		d, err := downstream.GetFromDeployToken(c.RequestURL().Query().Get("token"))
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to get cluster id from deploy token"))
+			return
+		}
+
+		logger.Info(fmt.Sprintf("cluster %s connected", d.ClusterID))
+		c.Join(d.ClusterID)
+
+		clusterSocket := ClusterSocket{
+			ClusterID:             d.ClusterID,
+			SocketID:              c.Id(),
+			SentPreflightURLs:     make(map[string]bool, 0),
+			LastDeployedSequences: make(map[string]int64, 0),
+		}
+		clusterSocketHistory = append(clusterSocketHistory, clusterSocket)
+	})
+
+	server.On(socket.OnDisconnection, func(c *socket.Channel) {
+		socketMtx.Lock()
+		defer socketMtx.Unlock()
+
+		updatedClusterSocketHistory := []ClusterSocket{}
+		for _, clusterSocket := range clusterSocketHistory {
+			if clusterSocket.SocketID != c.Id() {
+				updatedClusterSocketHistory = append(updatedClusterSocketHistory, clusterSocket)
+			}
+		}
+		clusterSocketHistory = updatedClusterSocketHistory
+	})
+
+	startLoop(deployLoop, 1)
+	startLoop(supportBundleLoop, 1)
+	startLoop(restoreLoop, 1)
+
+	return server
+}
+
+func startLoop(fn func(), intervalInSeconds time.Duration) {
+	go func() {
+		for {
+			fn()
+			time.Sleep(time.Second * intervalInSeconds)
+		}
+	}()
+}
+
+func deployLoop() {
+	for _, clusterSocket := range clusterSocketHistory {
+		apps, err := app.ListInstalledForDownstream(clusterSocket.ClusterID)
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to list installed apps for downstream"))
+			continue
+		}
+
+		for _, a := range apps {
```

Comment:
could you make everything inside here a function and then you could return rather than continue and get rid of a lot of logger.Error statements?